src/PROPOSAL/detail/PROPOSAL/Propagator.cxx

```diff
@@ -139,93 +142,112 @@ Interaction::Loss Propagator::DoStochasticInteraction(ParticleState& p_cond,
+        }
 
-            AdvanceDistance[ReachedBorder] = CalculateDistanceToBorder(
-                state.position, mean_direction, *geometry);
-            advancement_type = minimize(AdvanceDistance);
-            control_distance = AdvanceDistance[advancement_type];
-        } while (std::abs(advance_distance - control_distance)
-            > PARTICLE_POSITION_RESOLUTION);
-        advance_distance = control_distance;
-        advance_grammage = density->Calculate(
-            state.position, mean_direction, advance_distance);
-        new_position = state.position + advance_distance * mean_direction;
-    }
+        // Calculate scattering proposal
+        std::tie(mean_direction, new_direction) = utility.DirectionsScatter(
+                grammage, state.energy, energy, state.direction, rnd);
+
+        // Check step
+        double distance_to_border = CalculateDistanceToBorder(state.position, mean_direction, *geometry);
+        bool is_inside = geometry->IsInside(state.position, mean_direction);
+        if (!is_inside) {
+            // Special case: We are on the sector border, but scattering back outside the current sector!
+            // Update sector and recalculate values
+            advancement_type = InvalidStep;
+            auto new_sector = GetCurrentSector(state.position, mean_direction);
+            utility = get<UTILITY>(new_sector);
+            density = get<DENSITY_DISTR>(new_sector);
+            geometry = get<GEOMETRY>(new_sector);
+            grammage_next_interaction = utility.LengthContinuous(state.energy, energy_next_interaction);
+            energy = energy_next_interaction;
+            distance = -1;
+            grammage = -1;
+        } else if (distance <= distance_to_border && distance <= max_distance && energy == energy_next_interaction) {
+            // reached interaction
+            advancement_type = ReachedInteraction;
+        } else if (distance <= distance_to_border && distance == max_distance) {
+            // reached max distance
+            advancement_type = ReachedMaxDistance;
+        } else if (std::abs(distance - distance_to_border) <= PARTICLE_POSITION_RESOLUTION) {
+            // reached geometry border
+            advancement_type = ReachedBorder;
+            distance = distance_to_border;
+        } else {
+            // iteration not finished! discard energy and grammage
+            advancement_type = InvalidStep;
+            distance = std::min(distance_to_border, max_distance);
+            energy = -1;
+            grammage = -1;
+        }
+    } while (advancement_type == InvalidStep);
 
-    state.time = state.time
-        + utility.TimeElapsed(state.energy, E_f, advance_grammage,
-            density->Evaluate(state.position));
-    state.position = new_position;
+    state.time = state.time + utility.TimeElapsed(state.energy, energy, grammage, density->Evaluate(state.position)); // TODO: should the energy passed here be the randomized energy or not?
```

Comment:
This is quite a good point. Historically, the randomized energy has always been calculated after the propagation time.
The reason for this was, that the time and the distance were calculated together in the old `AdvanceParticle` method, see e.g. [here](https://github.com/tudo-astroparticlephysics/PROPOSAL/blob/v3.0.0/private/PROPOSAL/Propagator.cxx#L756).
Since then we have kept the order of calculation.
Thinking of it, the time calculation should depend on the randomized energy.
But this should be discussed in the next PROPOSAL meeting.

Reply:
If we decide to change the order here, I think we should add this in a separate PR. In my opinion, this is pretty significant, and should not be "hidden" in this restructuring PR.

So I would say that we merge it with the "old" order, and make a new PR where we only change the order of the continuous randomization calculation.

Reply:
ok